src/core/expr/fexpr_qcut.cc

```diff
@@ -0,0 +1,198 @@
+      if (ctx.has_groupby()) {
+        throw NotImplError() << "qcut() cannot be used in a groupby context";
+      }
+
+      Workframe wf = arg_->evaluate_n(ctx);
+      const size_t ncols = wf.ncols();
+
+      int32vec nquantiles(ncols);
+      bool defined_nquantiles = !py_nquantiles_.is_none();
+      bool nquantiles_list_or_tuple = py_nquantiles_.is_list_or_tuple();
+
+      if (nquantiles_list_or_tuple) {
+        py::oiter py_nquantiles = py_nquantiles_.to_oiter();
+        if (py_nquantiles.size() != ncols) {
+          throw ValueError() << "When `nquantiles` is a list or a tuple, its "
+            << "length must be the same as the number of input columns, i.e. `"
+            << ncols << "`, instead got: `" << py_nquantiles.size() << "`";
+
+        }
+
+        size_t i = 0;
+        for (auto py_nquantile : py_nquantiles) {
+          int32_t nquantile = py_nquantile.to_int32_strict();
+          if (nquantile <= 0) {
+            throw ValueError() << "All elements in `nquantiles` must be positive, "
+              << "got `nquantiles[" << i << "`]: `" << nquantile << "`";
+          }
+
+          nquantiles[i++] = nquantile;
+        }
+        xassert(i == ncols);
+      }
+      else {
+        int32_t nquantiles_default = 10;
+        if (defined_nquantiles) {
+          nquantiles_default = py_nquantiles_.to_int32_strict();
+          if (nquantiles_default <= 0) {
+            throw ValueError() << "Number of quantiles must be positive, "
+              "instead got: `" << nquantiles_default << "`";
+          }
+        }
+
+        for (size_t i = 0; i < ncols; ++i) {
+          nquantiles[i] = nquantiles_default;
+        }
+      }
+
+      // Qcut workframe in-place
+      for (size_t i = 0; i < ncols; ++i) {
+        Column coli = wf.retrieve_column(i);
+
+        if (coli.ltype() == dt::LType::STRING || coli.ltype() == dt::LType::OBJECT)
+        {
+          throw TypeError() << "`qcut()` cannot be applied to "
+            << "string or object columns, instead column `" << i
+            << "` has an stype: `" << coli.stype() << "`";
+        }
+
+        coli = Column(new Latent_ColumnImpl(new Qcut_ColumnImpl(
+                 std::move(coli), nquantiles[i]
+               )));
+        wf.replace_column(i, std::move(coli));
+      }
+
+      return wf;
+    }
+};
+
+
+
+
+//------------------------------------------------------------------------------
+// Python-facing `qcut()` function
+//------------------------------------------------------------------------------
+
+static const char* doc_qcut =
+R"(qcut(cols, nquantiles=10)
+--
+
+Bin all the columns from `cols` into intervals with approximately
+equal populations. Thus, the intervals are chosen according to
+the sample quantiles of the data.
+
+If there are duplicate values in the data, they will all be placed
+into the same been. In extreme cases this may cause the bins to be
+highly unbalanced.
+
+Parameters
+----------
+cols: FExpr
```

Comment:
I remember we discussed that the input can be also different from `FExpr`?

Reply:
Yeah, we'll change this (and other places) once we have a documentation for what it means to be "FExpr-like".